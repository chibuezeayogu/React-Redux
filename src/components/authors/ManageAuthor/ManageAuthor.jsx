import React, { Component, Fragment } from 'react';
import { connect } from 'react-redux';
import { withRouter, Prompt } from 'react-router-dom';

import { saveAuthor } from '../../../actions/authorActions';
import AuthorsForm from '../authorsForm/AuthorsForm';
import getById from '../../../helpers/getCourseById';
import inputValidation from '../../../helpers/inputValidation';


/**
 * @class ManageAuthors
 * @extends Component
 */
class ManageAuthor extends Component {
  static initialState = () => ({
  	author: {
  		id: '',
  		firstName: '',
  		lastName: ''
  	},
  	hasUnsavedChanges: false,
  	loading: false,
  	errors: {}
  })

  state = ManageAuthor.initialState();

  static getDerivedStateFromProps(nextProps, prevState) {
  	if (nextProps.author.id && !prevState.author.id) {
  		return {
  			...ManageAuthor.initialState(),
  			author: nextProps.author
  		};
  	}
    
  	return null;
  }
  
  /**
   * 
   * @memberof ManageAuthor
   * @param {Object} event
   * @returns {void}
   */
  onChange = (event) => {
  	const { errors, author } = this.state;
  
  	if (Object.entries(errors).length > 0) {
  		this.setState({ errors: {} });
  	}
  	author[event.target.name] = event.target.value;
    
  	this.setState({ 
  		author,
  		hasUnsavedChanges: true
  	});
  }
  
  /**
   * 
   * @memberof ManageAuthors
   * @param {Object} evt
   * @returns {void}
   */
  saveAuthor = (evt) => {
  	evt.preventDefault();
  	const err = inputValidation(this.state.author);
  	if (!err.isEmpty) {
  		this.setState({ errors: err.errors });
  	} else {
  		this.setState({ loading: true });
  		this.props.saveAuthor(this.state.author)
  			.then(() => {
  				this.setState({ 
  					hasUnsavedChanges: false,
  					loading: false
  				});
  				this.props.history.push('/authors');
  			});
  	}
  }
  
  /**
   * @memberof render
   * @return {void} JSX element
   */
  render() {
  	const {
  		author, hasUnsavedChanges, loading, errors 
  	} = this.state;
  	return (
  		<Fragment>
  			<Prompt
  				when={hasUnsavedChanges}
  				message="Your unsaved data will be lost"
  			/>
  			<div className="jumbotron">
  				<h1 className="display-4">Manage Author</h1>
  			</div>
  			<AuthorsForm
  				author={author}
  				onChange={this.onChange}
  				errors={errors}
  				onSave={this.saveAuthor}
  				loading={loading}
  			/>
  		</Fragment>
  	);
  }
}

const mapStateToProps = ({ coursesReducer, authorsReducer }, props) => {
	const { id } = props.match.params;
	let author = ManageAuthor.initialState();
  
	if (id && authorsReducer.authors.length > 0) {
		author = Object.assign({}, getById(authorsReducer.authors, id));
	}
  
	return {
		author,
		courses: coursesReducer.courses
	};
}; 

export default connect(mapStateToProps, { saveAuthor })(ManageAuthor);
