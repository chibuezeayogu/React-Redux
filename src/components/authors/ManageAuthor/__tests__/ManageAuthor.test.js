import React from 'react';
import { shallow } from 'enzyme';

import { ManageAuthor } from '../ManageAuthor';


describe('ManageAuthor Component', () => {
	const props = {
		history: {
			push: jest.fn()
		},
		saveAuthor: jest.fn().mockImplementation(() => Promise.resolve()),
		match: { params: { id: 'courseID' } },
		courses: [{
			id: 'courseID',
			watchHref: '',
			title: '',
			authorId: '',
			length: '',
			category: ''
		}],
		authors: [{
			id: 'cory-house',
			firstName: 'Cory',
			lastName: 'House'
		}]
	};

	const wrapper = shallow(<ManageAuthor {...props} />);

	it('should render correctly', () => {
		expect(wrapper).toMatchSnapshot();
		expect(wrapper.length).toEqual(1);
		expect(wrapper.find('div.jumbotron')).toHaveLength(1);
		expect(wrapper.find('h1.display-4')).toHaveLength(1);
		expect(wrapper.find('h1').text()).toBe('Manage Author');
	});

	it('should call onChange method', () => {
		const spy = jest.spyOn(wrapper.instance(), 'onChange');
		const event = {
			target: {
				name: 'firstName',
				value: 'firstName'
			}
		};
		wrapper.instance().onChange(event);
		expect(spy).toHaveBeenCalled();
	});

	it('should return an error message if firstName is empty', () => {
		const spy = jest.spyOn(wrapper.instance(), 'saveAuthor');
		wrapper.instance().saveAuthor({ preventDefault: jest.fn() });
		expect(spy).toHaveBeenCalled();
	});

	it('should saveAuthor if correct values are supplied', () => {
		const spy = jest.spyOn(wrapper.instance(), 'saveAuthor');
		wrapper.setState({
			author: {
				firstName: 'Chibueze',
				lastName: 'Ayogu',
			}
		});
		wrapper.instance().saveAuthor({ preventDefault: jest.fn() });
		expect(spy).toHaveBeenCalled();
	});
});