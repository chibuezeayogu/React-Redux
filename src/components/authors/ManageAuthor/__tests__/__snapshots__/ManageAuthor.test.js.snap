// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ManageAuthor Component should render correctly 1`] = `
<Fragment>
  <Prompt
    message="Your unsaved data will be lost"
    when={false}
  />
  <div
    className="jumbotron"
  >
    <h1
      className="display-4"
    >
      Manage Author
    </h1>
  </div>
  <AuthorForm
    author={Object {}}
    errors={Object {}}
    loading={false}
    onChange={[Function]}
    onSave={[Function]}
  />
</Fragment>
`;
